src/trio/_repl.py

```diff
@@ -42,7 +39,13 @@ async def _runcode_in_trio() -> outcome.Outcome[object]:
             # repl, but an error in the code. So we print the exception
             # and stay in the repl.
             raise
-        except BaseException:
+        except BaseException as exc:
+            # remove our code from traceback
+            tb = exc.__traceback__
+            for _ in range(n):
+                assert tb is not None
+                tb = tb.tb_next
+            exc.__traceback__ = tb
```

Comment:
I haven't put much thought into this so I might be wrong, but I think that using outcome would decrease the number of trio frames in all Python versions: this would mean nicer (but not optimally nice) tracebacks in pre-3.11 without messing around with tracebacks here.

Reply:
The general change proposed makes sense to me. However, getting the traceback neatly formatted doesn't look straightforward.

I had a quick play using `outcome`. This has the advantage that the sync and async traceback have the same number of frames that need to be remove. So no need to use the count variable to know how many levels to strip.

```python
    def runcode(self, code: types.CodeType) -> None:
        try:
            func = types.FunctionType(code, self.locals)
            if inspect.iscoroutinefunction(func):
                trio.from_thread.run(outcome.acapture, func).unwrap()
            else:
                trio.from_thread.run_sync(outcome.capture, func).unwrap()
        except SystemExit:
            raise
        except BaseException as exc:
            self.showtraceback()
```

The `InteractiveConsole.showtraceback()` already strips one level of the traceback. So I tried copying that into the `TrioInteractiveConsole` subclass and modifying it to strip two levels. I think this works. One complication is the original logic checks to see if `sys.excepthook` has been set and if so, doesn't strip the traceback frame. However in any python prior to 3.12, Trio includes `exceptiongroup` which *does* override the default `sys.excepthook`.

In the implementation below, I'm stripping the two stack frame even when delegating to `sys.excepthook`. That is a little different to the stdlib version, but perhaps is a pragmatic middle ground.

```python
    def showtraceback(self):
        sys.last_type, sys.last_value, last_tb = ei = sys.exc_info()
        sys.last_traceback = last_tb
        # see https://docs.python.org/3/library/sys.html#sys.last_exc
        if sys.version_info >= (3, 12):
            sys.last_exc = ei[1]  
        
        try:
            lines = traceback.format_exception(ei[0], ei[1], last_tb.tb_next.tb_next)
            if sys.excepthook is sys.__excepthook__:
                self.write(''.join(lines))
            else:
                # pre 3.12, exceptgroup installs its own hook, so this
                # path gets taken.
                # was -> sys.excepthook(ei[0], ei[1], last_tb)
                sys.excepthook(ei[0], ei[1], last_tb.tb_next.tb_next)
        finally:
            last_tb = ei = None
```

Reply:
As long as we are using outcome to clean up tracebacks, we can save further machinations by not raising and catching the error again, see fff91de.